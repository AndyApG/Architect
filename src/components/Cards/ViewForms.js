import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { map, has, get } from 'lodash';
import { TransitionGroup } from 'react-transition-group';
import { Node, Button } from '../../ui/components';
import { Wipe } from '../Transitions';
import { Guided } from '../Guided';
import Guidance from '../Guidance';
import Card from '../Card';
import { getProtocol } from '../../selectors/protocol';
import { actionCreators as formActions } from '../../ducks/modules/protocol/forms';

class ViewForms extends Component {
  handleDelete = (form) => {
    // eslint-disable-next-line no-alert
    if (confirm(`Are you sure you want to delete "${form}"?`)) {
      this.props.deleteForm(form);
    }
  };

  renderForm = (form, key) => (
    <Wipe key={key}>
      <div className="list__item" key={key}>
        <div className="list__attribute list__attribute--icon">
          <Link to={`${this.props.protocolPath}/form/${key}`}>
            <Node label={form.type} color={get(this.props.nodes, [form.type, 'color'], '')} />
          </Link>
        </div>
        <div className="list__attribute">
          <h3>
            <Link to={`${this.props.protocolPath}/form/${key}`}>
              {form.title}
            </Link>
          </h3>
        </div>
        <div className="list__attribute list__attribute--options">
          <Button size="small" color="neon-coral" onClick={() => this.handleDelete(key)}>
            Delete
          </Button>
        </div>
      </div>
    </Wipe>
  );

  renderForms() {
    const forms = this.props.forms;

    if (forms.length === 0) {
      return 'No forms defined';
    }

    return (
      <div className="list">
        <TransitionGroup>
          {map(forms, this.renderForm)}
        </TransitionGroup>
      </div>
    );
  }

  renderButtons() {
    return ([
      <Button key="cancel" size="small" color="platinum" onClick={this.props.onComplete}>Back</Button>,
    ]);
  }

  render() {
    const {
      show,
      protocolPath,
    } = this.props;

    return (
      <Card
        show={show}
        buttons={this.renderButtons()}
      >
        <div className="editor variable-registry">
          <Guided className="editor__sections">
            <h1>Form Manager</h1>

            <Guidance contentId="guidance.forms.index">
              <div className="editor__section">
                <h2>Existing Forms</h2>
                <div className="editor__subsection">
                  {this.renderForms()}
                </div>
                { protocolPath &&
                  <div className="editor__subsection">
                    <Link
                      to={`${protocolPath}/form/`}
                      className="button button--small"
                    >
                      Create new Form
                    </Link>
                  </div>
                }
              </div>
            </Guidance>
          </Guided>
        </div>
      </Card>
    );
  }
}

ViewForms.propTypes = {
  show: PropTypes.bool,
  forms: PropTypes.object.isRequired,
  nodes: PropTypes.object.isRequired,
  protocolPath: PropTypes.string,
  onComplete: PropTypes.func,
  deleteForm: PropTypes.func.isRequired,
};

ViewForms.defaultProps = {
  protocolPath: null,
  show: true,
  onComplete: () => {},
};

const mapStateToProps = (state, props) => {
  const protocol = getProtocol(state);

  return {
    forms: get(protocol, 'forms', {}),
    nodes: get(protocol, 'variableRegistry.node', {}),
    protocolPath: has(props, 'match.params.protocol') ?
      `/edit/${get(props, 'match.params.protocol')}` : null,
  };
};

const mapDispatchToProps = dispatch => ({
  deleteForm: bindActionCreators(formActions.deleteForm, dispatch),
});

export { ViewForms };

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ViewForms);
