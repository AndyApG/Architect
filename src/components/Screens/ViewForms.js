import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { map, get } from 'lodash';
import { TransitionGroup } from 'react-transition-group';
import { Node, Button } from '../../ui/components';
import { Wipe } from '../Transitions';
import { Guided } from '../Guided';
import Guidance from '../Guidance';
import Card from '../Card';
import Link from '../Link';
import { getProtocol } from '../../selectors/protocol';
import { actionCreators as formActions } from '../../ducks/modules/protocol/forms';
import { actionCreators as dialogsActions } from '../../ducks/modules/dialogs';

class ViewForms extends Component {
  get buttons() {
    return ([
      <Button key="cancel" color="platinum" onClick={this.handleCancel}>Back</Button>,
    ]);
  }

  handleDelete = (form) => {
    const formTitle = this.props.forms[form].title;

    this.props.openDialog({
      type: 'Warning',
      title: `Delete "${formTitle}" form`,
      message: `Are you sure you want to delete "${formTitle}"?`,
      onConfirm: () => { this.props.deleteForm(form); },
      confirmLabel: 'Delete form',
    });
  };

  handleCancel = this.props.onComplete;

  renderForm = (form, id) => (
    <Wipe key={id}>
      <div className="simple-list__item">
        <div className="simple-list__attribute simple-list__attribute--icon">
          <Link
            screen="form"
            params={{ id }}
          >
            <Node label={get(this.props.nodes, [form.type, 'label'], '')} color={get(this.props.nodes, [form.type, 'color'], '')} />
          </Link>
        </div>
        <div className="simple-list__attribute">
          <h3>
            <Link
              screen="form"
              params={{ id }}
            >
              {form.title}
            </Link>
          </h3>
        </div>
        <div className="simple-list__attribute simple-list__attribute--options">
          <Button size="small" color="neon-coral" onClick={() => this.handleDelete(id)}>
            Delete
          </Button>
        </div>
      </div>
    </Wipe>
  );

  renderForms() {
    const forms = this.props.forms;

    if (forms.length === 0) {
      return 'No forms defined';
    }

    return (
      <div className="simple-list">
        <TransitionGroup>
          {map(forms, this.renderForm)}
        </TransitionGroup>
      </div>
    );
  }

  render() {
    const {
      show,
      state,
    } = this.props;

    return (
      <Card
        show={show}
        state={state}
        buttons={this.buttons}
      >
        <Guided>
          <div className="editor variable-registry">
            <div className="editor__window">
              <div className="editor__content">
                <h1 className="editor__heading">Form Manager</h1>
                <p>
                  Use this screen to create, edit, and manage your forms.
                </p>
                <Guidance contentId="guidance.forms.index">
                  <div className="editor__section">
                    <h2>Existing Forms</h2>
                    <div className="editor__subsection">
                      {this.renderForms()}
                    </div>
                    <div className="editor__subsection">
                      <Link
                        screen="form"
                      >
                        <Button>
                          Create new Form
                        </Button>
                      </Link>
                    </div>
                  </div>
                </Guidance>
              </div>
            </div>
          </div>
        </Guided>
      </Card>
    );
  }
}

ViewForms.propTypes = {
  show: PropTypes.bool,
  state: PropTypes.object,
  forms: PropTypes.object.isRequired,
  nodes: PropTypes.object.isRequired,
  onComplete: PropTypes.func,
  deleteForm: PropTypes.func.isRequired,
  openDialog: PropTypes.func.isRequired,
};

ViewForms.defaultProps = {
  show: true,
  state: null,
  onComplete: () => {},
};

const mapStateToProps = (state) => {
  const protocol = getProtocol(state);

  return {
    forms: get(protocol, 'forms', {}),
    nodes: get(protocol, 'variableRegistry.node', {}),
  };
};

const mapDispatchToProps = dispatch => ({
  deleteForm: bindActionCreators(formActions.deleteForm, dispatch),
  openDialog: bindActionCreators(dialogsActions.openDialog, dispatch),
});

export { ViewForms };

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ViewForms);
