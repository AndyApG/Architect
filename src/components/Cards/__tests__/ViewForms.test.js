/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { ViewForms } from '../ViewForms';

const mockProps = {
  forms: {},
  nodes: {},
  deleteForm: () => {},
};

describe('<ViewForms />', () => {
  it('can render', () => {
    const component = shallow(<ViewForms {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
