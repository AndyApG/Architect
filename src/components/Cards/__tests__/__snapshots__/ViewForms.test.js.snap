// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewForms /> can render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ViewForms
    deleteForm={[Function]}
    forms={Object {}}
    nodes={Object {}}
    onComplete={[Function]}
    protocolPath={null}
    show={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "buttons": Array [
        <Button
          color="platinum"
          content=""
          icon=""
          iconPosition="right"
          onClick={[Function]}
          size="small"
        >
          Back
        </Button>,
      ],
      "children": <div
        className="editor variable-registry"
      >
        <Connect(Guided)
          className="editor__sections"
        >
          <h1>
            Form Manager
          </h1>
          <Connect(Guided)
            contentId="guidance.forms.index"
          >
            <div
              className="editor__section"
            >
              <h2>
                Existing Forms
              </h2>
              <div
                className="editor__subsection"
              >
                <div
                  className="list"
                >
                  <TransitionGroup
                    childFactory={[Function]}
                    component="div"
                  />
                </div>
              </div>
            </div>
          </Connect(Guided)>
        </Connect(Guided)>
      </div>,
      "show": true,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Connect(Guided)
          className="editor__sections"
        >
          <h1>
            Form Manager
          </h1>
          <Connect(Guided)
            contentId="guidance.forms.index"
          >
            <div
              className="editor__section"
            >
              <h2>
                Existing Forms
              </h2>
              <div
                className="editor__subsection"
              >
                <div
                  className="list"
                >
                  <TransitionGroup
                    childFactory={[Function]}
                    component="div"
                  />
                </div>
              </div>
            </div>
          </Connect(Guided)>
        </Connect(Guided)>,
        "className": "editor variable-registry",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <h1>
              Form Manager
            </h1>,
            <Connect(Guided)
              contentId="guidance.forms.index"
            >
              <div
                className="editor__section"
              >
                <h2>
                  Existing Forms
                </h2>
                <div
                  className="editor__subsection"
                >
                  <div
                    className="list"
                  >
                    <TransitionGroup
                      childFactory={[Function]}
                      component="div"
                    />
                  </div>
                </div>
              </div>
            </Connect(Guided)>,
          ],
          "className": "editor__sections",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Form Manager",
            },
            "ref": null,
            "rendered": "Form Manager",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <div
                className="editor__section"
              >
                <h2>
                  Existing Forms
                </h2>
                <div
                  className="editor__subsection"
                >
                  <div
                    className="list"
                  >
                    <TransitionGroup
                      childFactory={[Function]}
                      component="div"
                    />
                  </div>
                </div>
              </div>,
              "contentId": "guidance.forms.index",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <h2>
                    Existing Forms
                  </h2>,
                  <div
                    className="editor__subsection"
                  >
                    <div
                      className="list"
                    >
                      <TransitionGroup
                        childFactory={[Function]}
                        component="div"
                      />
                    </div>
                  </div>,
                  null,
                ],
                "className": "editor__section",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Existing Forms",
                  },
                  "ref": null,
                  "rendered": "Existing Forms",
                  "type": "h2",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <div
                      className="list"
                    >
                      <TransitionGroup
                        childFactory={[Function]}
                        component="div"
                      />
                    </div>,
                    "className": "editor__subsection",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <TransitionGroup
                        childFactory={[Function]}
                        component="div"
                      />,
                      "className": "list",
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "childFactory": [Function],
                        "children": Array [],
                        "component": "div",
                      },
                      "ref": null,
                      "rendered": Array [],
                      "type": [Function],
                    },
                    "type": "div",
                  },
                  "type": "div",
                },
                null,
              ],
              "type": "div",
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "buttons": Array [
          <Button
            color="platinum"
            content=""
            icon=""
            iconPosition="right"
            onClick={[Function]}
            size="small"
          >
            Back
          </Button>,
        ],
        "children": <div
          className="editor variable-registry"
        >
          <Connect(Guided)
            className="editor__sections"
          >
            <h1>
              Form Manager
            </h1>
            <Connect(Guided)
              contentId="guidance.forms.index"
            >
              <div
                className="editor__section"
              >
                <h2>
                  Existing Forms
                </h2>
                <div
                  className="editor__subsection"
                >
                  <div
                    className="list"
                  >
                    <TransitionGroup
                      childFactory={[Function]}
                      component="div"
                    />
                  </div>
                </div>
              </div>
            </Connect(Guided)>
          </Connect(Guided)>
        </div>,
        "show": true,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Connect(Guided)
            className="editor__sections"
          >
            <h1>
              Form Manager
            </h1>
            <Connect(Guided)
              contentId="guidance.forms.index"
            >
              <div
                className="editor__section"
              >
                <h2>
                  Existing Forms
                </h2>
                <div
                  className="editor__subsection"
                >
                  <div
                    className="list"
                  >
                    <TransitionGroup
                      childFactory={[Function]}
                      component="div"
                    />
                  </div>
                </div>
              </div>
            </Connect(Guided)>
          </Connect(Guided)>,
          "className": "editor variable-registry",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <h1>
                Form Manager
              </h1>,
              <Connect(Guided)
                contentId="guidance.forms.index"
              >
                <div
                  className="editor__section"
                >
                  <h2>
                    Existing Forms
                  </h2>
                  <div
                    className="editor__subsection"
                  >
                    <div
                      className="list"
                    >
                      <TransitionGroup
                        childFactory={[Function]}
                        component="div"
                      />
                    </div>
                  </div>
                </div>
              </Connect(Guided)>,
            ],
            "className": "editor__sections",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Form Manager",
              },
              "ref": null,
              "rendered": "Form Manager",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <div
                  className="editor__section"
                >
                  <h2>
                    Existing Forms
                  </h2>
                  <div
                    className="editor__subsection"
                  >
                    <div
                      className="list"
                    >
                      <TransitionGroup
                        childFactory={[Function]}
                        component="div"
                      />
                    </div>
                  </div>
                </div>,
                "contentId": "guidance.forms.index",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <h2>
                      Existing Forms
                    </h2>,
                    <div
                      className="editor__subsection"
                    >
                      <div
                        className="list"
                      >
                        <TransitionGroup
                          childFactory={[Function]}
                          component="div"
                        />
                      </div>
                    </div>,
                    null,
                  ],
                  "className": "editor__section",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Existing Forms",
                    },
                    "ref": null,
                    "rendered": "Existing Forms",
                    "type": "h2",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <div
                        className="list"
                      >
                        <TransitionGroup
                          childFactory={[Function]}
                          component="div"
                        />
                      </div>,
                      "className": "editor__subsection",
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": <TransitionGroup
                          childFactory={[Function]}
                          component="div"
                        />,
                        "className": "list",
                      },
                      "ref": null,
                      "rendered": Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "class",
                        "props": Object {
                          "childFactory": [Function],
                          "children": Array [],
                          "component": "div",
                        },
                        "ref": null,
                        "rendered": Array [],
                        "type": [Function],
                      },
                      "type": "div",
                    },
                    "type": "div",
                  },
                  null,
                ],
                "type": "div",
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
